File: benchmarks/blur/python/gaussian_blur.py
from typing import Optional
import math
from lib.python.profiler import Profiler
from benchmarks.blur.python.image_util import ImageUtil


class GaussianBlur:
    def __init__(self, kernel_radius: int):
        self.kernel_radius: int = kernel_radius
        self.kernel_size: int = 2 * kernel_radius + 1
        self.kernel: Optional[list[float]] = None

        self.width: Optional[int] = None
        self.height: Optional[int] = None

    def _create_kernel(self) -> list[float]:
        """
        Create a 1D Gaussian kernel with kernel_radius

        Returns:
            kernel: A 1D float array of the kernel
        """

        self.kernel = [0.0] * self.kernel_size
        std = self.kernel_size / 6

        self.kernel = [1 / (math.sqrt(2 * math.pi) * std) *
                       math.exp(-(i - self.kernel_radius)**2 / (2 * std**2)) for i in range(self.kernel_size)]

        return self.kernel

    def _horizontal_blur(self, padded: list[list[float]]) -> list[list[float]]:
        """
        Apply blur horizontally

        Args:
            padded: Padded image
        Returns:
            h_blurred: Resulting image array after blurring horizontally
        """

        pad_size = self.kernel_radius
        padded_height = len(padded)
        padded_width = len(padded[0])

        # clone padded first to preserve paddings
        h_blurred = [row[:] for row in padded]

        for y in range(padded_height):
            for x in range(pad_size * 3, padded_width - pad_size * 3, 3):
                r = 0
                g = 0
                b = 0

                for i in range(self.kernel_size):
                    r = r + self.kernel[i] * padded[y][x -
                                                       self.kernel_radius * 3 + i * 3]
                    g = g + self.kernel[i] * padded[y][x -
                                                       self.kernel_radius * 3 + i * 3 + 1]
                    b = b + self.kernel[i] * padded[y][x -
                                                       self.kernel_radius * 3 + i * 3 + 2]

                h_blurred[y][x] = r
                h_blurred[y][x + 1] = g
                h_blurred[y][x + 2] = b

            print(
                f"\rHorizontal blur: {100 * y / padded_height:.2f}%", end="", flush=True)

        print("\r\033[k", end="", flush=True)

        return h_blurred

    def _vertical_blur(self, h_blurred: list[list[float]]) -> list[list[float]]:
        """
        Apply blur vertically

        Args:
            h_blurred Horizontally blurred image (therefore should run horizontal_blur first)
        Returns:
            blurred: Resulting image array after blurring
        """

        pad_size = self.kernel_radius
        padded_height = len(h_blurred)
        padded_width = len(h_blurred[0])

        blurred = [[0.0] * self.width] * self.height

        for x in range(pad_size * 3, padded_width - pad_size * 3):
            for y in range(pad_size, padded_height - pad_size):
                c = sum(self.kernel[i] * h_blurred[y - self.kernel_radius + i][x]
                        for i in range(self.kernel_size))

                blurred[y - pad_size][x-pad_size * 3] = c

            print(
                f"\rVertical blur: {100 * (x - pad_size * 3) / self.width:.2f}%", end="", flush=True)

        print("\r\033[k", end="", flush=True)

        return blurred

    def blur(self, input_path: str, output_path: str) -> str:
        """
        Performs Gaussian blur over the given image with given kernal radius. Using Gaussian blur's separability, this method runs horizontal blur first and applies vertical blur later, which is more efficient than applying a 2D Gaussian kernel.

        Args:
            input_path: Path to the input image relative to root
            output_path: Desired path of the output image relative to root.
        Returns:
            A table (string) of memory used and elapsed time for horizontal blur, vertical blur, and total time elapsed
        """

        profiler = Profiler()
        profiler.start("Total")

        pad_size = self.kernel_radius
        self._create_kernel()

        img = ImageUtil.load_image(input_path)
        padded = ImageUtil.pad_image(img, pad_size)

        self.height = len(img)
        self.width = len(img[0])

        profiler.start("Horizontal blur")
        h_blurred = self._horizontal_blur(padded)
        profiler.end("Horizontal blur")

        profiler.start("Vertical blur")
        blurred = self._vertical_blur(h_blurred)
        profiler.end("Vertical blur")

        ImageUtil.save_image(blurred, output_path)

        profiler.end("Total")

        return profiler.get_table()

File: benchmarks/blur/java/GaussianBlur.java
package benchmarks.blur.java;

import lib.java.Profiler;

class GaussianBlur {
    int kernelRadius;
    int kernelSize;
    double[] kernel;

    int width;
    int height;

    public GaussianBlur(int kernelRadius) {
        this.kernelRadius = kernelRadius;
        this.kernelSize = 2 * kernelRadius + 1;
    }

    /**
     * Create a 1D Gaussian kernel with kernelRadius
     * 
     * @return A 1D double array of the kernel
     */
    private double[] createKernel() {
        this.kernel = new double[kernelSize];
        double std = this.kernelSize / 6.0;

        for (int i = 0; i < kernelSize; i++) {
            int xFromCenter = i - this.kernelRadius;
            kernel[i] = 1 / (Math.sqrt(2 * Math.PI) * std) * Math.exp(-(xFromCenter * xFromCenter) / (2 * std * std));
        }

        return kernel;
    }

    /**
     * Apply blur horizontally
     * 
     * @param padded Padded image
     * @return Resulting image array after blurring horizontally
     */
    private double[][] horizontalBlur(double[][] padded) {
        int padSize = this.kernelRadius;
        int paddedHeight = padded.length;
        int paddedWidth = padded[0].length;

        double[][] hBlurred = new double[paddedHeight][paddedWidth];

        // Clone padded first to preserve paddings
        for (int y = 0; y < paddedHeight; y++) {
            for (int x = 0; x < paddedWidth; x++) {
                hBlurred[y][x] = padded[y][x];
            }
        }

        for (int y = 0; y < paddedHeight; y++) {
            for (int x = padSize * 3; x < paddedWidth - padSize * 3; x += 3) {
                double r = 0, g = 0, b = 0;

                for (int i = 0; i < this.kernelSize; i++) {
                    r += this.kernel[i] * padded[y][x - this.kernelRadius * 3 + i * 3];
                    g += this.kernel[i] * padded[y][x - this.kernelRadius * 3 + i * 3 + 1];
                    b += this.kernel[i] * padded[y][x - this.kernelRadius * 3 + i * 3 + 2];
                }

                hBlurred[y][x] = r;
                hBlurred[y][x + 1] = g;
                hBlurred[y][x + 2] = b;
            }
            System.out.print("\rHorizontal blur: " + String.format("%.2f", 100.0 * y / paddedHeight) + "%");
        }
        System.out.print("\r\033[K");

        return hBlurred;
    }

    /**
     * Apply blur vertically
     * 
     * @param hBlurred Horizontally blurred image (therefore should run horizontalBlur first)
     * @return Resulting image array after blurring
     */
    private double[][] verticalBlur(double[][] hBlurred) {
        int padSize = this.kernelRadius;
        int paddedHeight = hBlurred.length;
        int paddedWidth = hBlurred[0].length;

        double[][] blurred = new double[height][width]; // same size and type as original image

        for (int x = padSize * 3; x < paddedWidth - padSize * 3; x++) {
            for (int y = padSize; y < paddedHeight - padSize; y++) {
                double c = 0;

                for (int i = 0; i < this.kernelSize; i++) {
                    c += this.kernel[i] * hBlurred[y - this.kernelRadius + i][x];
                }

                blurred[y - padSize][x - padSize * 3] = c;
            }
            System.out.print("\rVertical blur: " + String.format("%.2f", 100.0 * (x - padSize * 3) / this.width) + "%");
        }
        System.out.print("\r\033[K");

        return blurred;
    }

    /**
     * Performs Gaussian blur over the given image with given kernal radius. Using
     * Gaussian blur's separability, this method runs horizontal blur first and
     * applices vertical blur later, which is more efficient thab applying a 2D
     * Gaussian kernel.
     * 
     * @param inputPath Path to the input image relative to root
     * @param outputPath Desired path of the output image relative to root
     * @return A table (string) of memory used and elapsed time for horizontal blur, vertical blur, and
     *         total time elapsed
     */
    public String blur(String inputPath, String outputPath) {
        Profiler profiler = new Profiler();
        profiler.start("Total");

        int padSize = kernelRadius;
        createKernel();

        int[][] img = ImageUtil.loadImage(inputPath);
        double[][] padded = ImageUtil.padImage(img, padSize);

        this.height = img.length;
        this.width = img[0].length;

        profiler.start("Horizontal Blur");
        double[][] hBlurred = horizontalBlur(padded);
        profiler.end("Horizontal Blur");

        profiler.start("Vertical Blur");
        double[][] blurred = verticalBlur(hBlurred);
        profiler.end("Vertical Blur");

        ImageUtil.saveImage(blurred, outputPath);

        profiler.end("Total");

        return profiler.getTable();
    }

}